Remove implicit seekable requirement from QPY 16.

kevinhartman opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
Removes the implicit requirement that the `file_obj` passed during `qpy.dump` be seekable accidentally introduced in: 

- #14571 

### Details and comments
As suggested in #14775, we use seeking only if it's supported by the input stream and otherwise fall back on creating a temporary `io.BytesIO` buffer for the circuits payload.

Fixes #14775

kevinhartman committed:
Remove implicit seekable requirement from QPY 16.

kevinhartman commented:
This still needs some form of testing to exercise both paths (the stream being seekable vs not).

I'm currently considering adding some logic to `test_qpy.py`'s `generate_qpy` function to perform generation with a non-seekable in memory buffer in addition to the file stream we already generate, and then comparing those. Curious if anyone has a better idea.

I'm also currently just dumping the entire circuits buffer into the file for the non-seekable case, which perhaps should be chunked in some way instead.

Cryoris commented on qiskit/qpy/common.py:
```diff
@@ -190,7 +190,7 @@ def write_type_key(file_obj, type_key):
         file_obj (File): A file like object that contains the QPY binary data.
         type_key (bytes): Type key to write.
```
Could you document the new return value?

kevinhartman replied:
Done in e2b2914

Cryoris reviewed (commented):
This LGTM, but the tests would be good as you mentioned. Couldn't we just add some roundtrip tests to `test/python/qpy/test_circuit_load_from_qpy.py` with a custom non-seekable buffer? Adding a lot of tests via `generate_qpy` seems overkill since we mainly want to test the non-seekable path instead of the format itself.

kevinhartman committed:
Bug fixes for unseekable path, plus chunked writing of circuit payload.

kevinhartman committed:
Make sure QPY tests don't fail once we bump QPY_COMPATIBILITY_VERSION.

kevinhartman committed:
Add testing for QPY unseekable streams.

kevinhartman committed:
Make write_type_key in QPY return the num bytes written.

kevinhartman commented:
@Cryoris 

> Couldn't we just add some roundtrip tests to test/python/qpy/test_circuit_load_from_qpy.py with a custom non-seekable buffer? 

Good idea. Done in 7c5368f. Good thing I added testing, there were a couple of bugs 🙂.

Cryoris commented on qiskit/qpy/interface.py:
```diff
@@ -198,40 +200,66 @@ def dump(
-    file_obj.write(header)
-    common.write_type_key(file_obj, type_keys.Program.CIRCUIT)
+    header_bytes_written = file_obj.write(header)
+    header_bytes_written += common.write_type_key(file_obj, type_keys.Program.CIRCUIT)
```
I've struggled to find precise docs on the requirements of `write`, but it seems that sometimes it can return `None` (see https://docs.python.org/3/library/io.html#io.RawIOBase.write for example). For us this would indicate an error, but it would be nice if we could catch that and raise a proper error.

Cryoris reviewed (dismissed):
LGTM I also verified that Jake's original test works (with the change that `write` actually does something). I left one optional comment below which I think would be nice to add, but it fixes the bug as is.

kevinhartman committed:
Don't rely on write to return bytes.

Instead, we now have a formats entry for TYPE_KEY and compute the
size statically.
